### DISCUSSION TITLE
Question about the function retrieve()

### DISCUSSION THREAD

**Discussion by BishopDrunkAgain** [Cyfrin/foundry-full-course-cu#746] [Q&A] [answered]
Hello guys,
I am very new to coding and I was confused by the function retrieve()  :

contract SimpleStorage {
 
        uint256 favoriteNumber;
        
        function store(uint _favoriteNumber) public {
            favoriteNumber = _favoriteNumber;
        }
        // view, pure functions
        function retrieve() public view returns(uint256) {
            return favoriteNumber;
        }
        }

I am confused with this exact part = 'public view returns (uint256)' , why we would add this part normally the 'return favoriteNumber' at the end is sufficient since favoriteNumber is already declared uint, 
Thank you in advance,
Kind regards.

**Comment by theirrationalone** [comment]
Hey @BishopDrunkAgain !,

As you may know, the `retrieve` function is a helper function...

There is a term in programming for these kinds of `functions` known as stateless, dumb, helper functions. These types of functions typically do not modify any state, such as variable values; instead, they are used to retrieve values from the code. They cannot modify or update variable values but can only access and read them.

**Public**: The `public` keyword is used to restrict the visibility of a function. If you include the `public` keyword with your function, the function becomes visible to anyone and can be accessed from anywhere, either from outside or inside the contract. These keywords are often referred to as `Visibility modifiers`, and Solidity has four of them: `public`, `private`, `internal`, and `external`. If you want to learn more about them, please check out the Solidity official documentation.

**View**: The `view` keyword is used to restrict the accessibility of a variable (state/constant/immutable/memory). If you include the `view` keyword with your function, the function can only access variables but cannot modify them. Another thing to keep in mind while working with the `view` keyword is that it's good practice to access state and immutable variables with it. On the other hand, memory and constant variables should be accessed with the `pure` keyword. In Solidity, `view` and `pure` keywords are also known as `Accessibility Modifiers`.

**Returns**: In Solidity, we need to specify the type of value a function will return, similar to some well-known programming languages like C, C++, Java, etc. If you add `returns(uint256)` to a function, it means that the function must return a value of type `uint256`."

**Comment by BishopDrunkAgain** [comment]
Thank you so much @theirrationalone ,
But what I still don't understand is, why we need to specify one more time the type of the variable, we already declared it when we typed :
uint256 favoriteNumber;
What got me confused is why the system doesn't read that the variable is of type uint256 why do we have to specify it twice, I understand if we had 2 variables of different types names favoriteNumber but since we have only one variable we shouldn't specify it since the system will go and read the variable and find it from uint256;
Thank you in advance,

**Comment by BishopDrunkAgain** [comment]
I mean, is it just a syntax and must be written like that? what's its purpose?

**Reply by theirrationalone** [reply, marked as answer]
Exactly as you said the system... It's solidity design paradigm... It's a syntax ... through syntax we tell functions what to return, in advance and it's must to apply. Since we're using precompiled language so purpose of each syntax in already defined into official docs although i also described  those syntax purposes into earlier comment.

The system could not read the return type for us in advance unlike as in `Javascript` or `Python`. So We must to follow solidity syntax patterns if we wanna work with it.

**Reply by BishopDrunkAgain** [reply]
@theirrationalone Thank you so much I perfectly understand now 👍

**Reply by theirrationalone** [reply]
You're most Welcome @BishopDrunkAgain ! 😉
